File: 1-arrays/cpp/1-two-sum.cpp
#include<bits/stdc++.h>
using namespace std;
vector<int> twoSum(vector<int>& nums, int target) {
    map<int, int> mp;

    for(int i=0; i<nums.size(); i++){
        if(mp.find(target - nums[i]) != mp.end()){
            return {mp[target - nums[i]], i};
        }
        mp.insert({nums[i], i});
    }
    return {};
}

int main() {
    vector<int> nums = {2,7,11,15};
    int target = 9;
    vector<int> ans;
    ans = twoSum(nums, target);

    for(auto i: ans){
        cout<< i<< " ";
    }
    cout<< endl;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 1-arrays/java/1-two-sum.java
import java.util.Arrays;
import java.util.Hashtable;

public class Solution{
    public static int[] twoSum(int[] nums, int target) {
        Hashtable<Integer, Integer> ht = new Hashtable<>();

        for(int i=0; i<nums.length; i++){
            if(ht.containsKey(target - nums[i])){
                return new int[]{ht.get(target - nums[i]), i};
            }
            ht.put(nums[i], i);
        }
        return null;
    }

    public static void main(String args[]){
        int nums[] = {2,7,11,15};
        int target = 9;
        System.out.println(Arrays.toString(twoSum(nums, target)));
    }
}

// Time Complexity: O(n)
// Space Complexity: O(n)
